### DISCUSSION TITLE
Incorrect `root_path` Handling Causes Duplicated Path Prefixes in FastAPI Deployments

### DISCUSSION THREAD

**Discussion by realchandan** [fastapi/fastapi#11977] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.get("/ping")
def ping() -> str:
    return "pong"
```

### Description

The bug involves FastAPI incorrectly applying the `root_path` twice when deployed behind proxies or within Docker containers. This results in duplicated path prefixes, leading to incorrect routing and broken API documentation URLs. I've created a minimal reproducible example that demonstrates the issue, showing how paths like `/api/api/ping` are generated instead of the correct `/api/ping`. This issue needs to be addressed to ensure proper routing in various deployment scenarios.

I've created a [Minimal Reproducible Example](https://github.com/realchandan/root-path-bug-mre) with traefik as reverse proxy.

I've found similar issues -

1. [https://github.com/fastapi/fastapi/issues/2872](https://github.com/fastapi/fastapi/issues/2872)
2. [https://github.com/fastapi/fastapi/discussions/9018](https://github.com/fastapi/fastapi/discussions/9018)

But they were ignored, thus I am creating this new one with a Minimal Reproducible Example and nobody can deny the existence of the bug anymore.

### Operating System

Windows

### FastAPI Version

0.112.0

### Pydantic Version

2.8.2

### Python Version

Python 3.12.4

**Comment by fbs** [comment, marked as answer]
I found that if you set `root_path` in uvicorn it updates the `path` in asgi, whereas setting `root_path` in the fastapi constructor doesn't do that (but breaks mounted app). The bottom of my question has the uvicorn behaviour as well: https://github.com/fastapi/fastapi/discussions/11959#discussioncomment-10261507

So in this case,can you try:

1. remove `--root-path` from the dockerfile
2. set `root_path` in `FastAPI()`

**Reply by realchandan** [reply]
That worked! Thank You @fbs !

**Reply by fbs** [reply]
it works but the root path situation is still quite confusing and the docs don't help either. So some traction on the issues would still be nice
